superset/assets/src/explore/controlUtils.js

```diff
@@ -0,0 +1,124 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+import controlPanelConfigs, { sectionsToRender } from './controlPanels';
+import controls from './controls';
+
+export function getFormDataFromControls(controlsState) {
+  const formData = {};
+  Object.keys(controlsState).forEach((controlName) => {
+    formData[controlName] = controlsState[controlName].value;
+  });
+  return formData;
+}
+
+export function validateControl(control) {
+  const validators = control.validators;
+  if (validators && validators.length > 0) {
+    const validatedControl = { ...control };
+    const validationErrors = [];
+    validators.forEach((f) => {
+      const v = f(control.value);
+      if (v) {
+        validationErrors.push(v);
+      }
+    });
+    delete validatedControl.validators;
+    return { ...validatedControl, validationErrors };
+  }
+  return control;
+}
+
+export function getControlKeys(vizType, datasourceType) {
+  const controlKeys = [];
+  sectionsToRender(vizType, datasourceType).forEach(
+    section => section.controlSetRows.forEach(
+      fsr => fsr.forEach(
```

Comment:
nit: can we give `fsr` and `f` more descriptive names? (I realize this was just moved over but I think it could help readability)